src/main/clojure/clara/rules/compiler.clj

```diff
@@ -855,11 +854,30 @@
 (defn- non-equality-unifications
   "Returns a set of unifications that do not use equality-based checks."
   [constraints]
-  (let [[bound-variables unbound-variables] (classify-variables constraints)]
-    (into #{}
-          (for [constraint constraints
-                :when (non-equality-unification? constraint bound-variables)]
-            constraint))))
+  (let [[bound-variables _] (classify-variables constraints)]
+    (loop [[cur & more] constraints
+           constraints #{}
+           bound-variables bound-variables]
+      (if cur
+        (let [non-equity? (non-equality-unification? cur bound-variables)
+
+              ;; In the event that the unification is also a binding, then we will need to remove
+              ;; the bound variable from the previously bound variables. This prevents further constraints
+              ;; that use this variable from being considered "equality-unifications".
+              ;; See https://github.com/cerner/clara-rules/issues/426 for more info.
+              bound-variables (if (and non-equity?
```

Comment:
I struggled with this a bit. I will summarize my understanding for clarity.

The issue #426  description had an example with this form of 
```clj
  [A (= ?a a)]
  [B (= ?r (conj b ?a))
     (not-empty? ?r)]
```
Now consider the constraints for `B` fact condition:
```clj
(= ?r (conj b ?a))
(not-empty? ?r)
```
What is happening is that the `(= ?r (conj b ?a))` is true for `(non-equality-unification? cur bound-variables)` because it has the sub-expression `(conj b ?a)` which is unifying on an unknown to this condition variable, in a way that is not using `=` (which has a simplified/optimized/hash-based impl).

I think it is important to note that this particular fn , `non-equality-unifications` is used to find all constraints that will be separated away from the "simple" expressions and moved into a "join expression" that deals with the non-equality unification.

Now `(classify-variables constraints)` is classifying this new variable binding `?r` as a "bound variable" in `bound-variables`. However, it isn't necessarily "previously bound" if constraints are moved around during compilation. Subsequent constraints that _are not_ moved with the non-equality unifications, could be broken, since the classification isn't accurate anymore if they are not also moved.

Concretely, the next constraint is `(not-empty? ?r)` and `?r` is classified as previously bound now.
This means that:
```clj
(non-equality-unification? '(not-empty? ?r) #{?r}) 
;;= false
```

This would result in `(not-empty? ?r)` being "skipped" in this fn, and consequently not "moved" with the rest of the "join filter" expression constraints later.

This approach works around this, by doing what Ethan appropriately called "re-classification". When a non-equality unification is detected, it has to remove it's newly introduced bindings from the classification as a mechanism to make other dependent constraints move along with it.

I can follow this approach. However, it does seem like a sneaky way to utilize how bound variables are used to detect non-equality unifications.
The reason I find it "sneaky" is that the logic it is doing is somewhat assuming that the caller is planning to move constraints around. It has the semantics that if it returns a constraint, it needs to also return all the constraints that may depend on that constraint's new bindings.

-----

However, it seems that this may _not_ work for some cases still:

```clj
(defrecord A [a])
(defrecord B [b])

(defrule a-rule2
  [A (= ?a a)]
  [B (= ?r (conj b ?a))
     (= ?r #{:a :b})]
  =>
  (println "doesn't matter"))
```

ie. the 2nd constraint is an equality binding on `?r`.

This compiles, but it isn't producing a semantically correct network.

The equality unification on `?r` in `(= ?r #{:a :b})` isn't being considered dependent on the `?r` introduced by `(= ?r (conj b ?a))` that is going to be moved into the join filter later.
You can see this by running above with:

```clj
(-> (mk-session [a-rule2] :cache false)
    (insert (->A :c)
            (->B #{:b}))
    fire-rules)

;; Prints "doesn't matter"

```

I think the solution to this is in not trying to trying to manipulate only the classified `bound-variables` to find the dependent constraints. A more "brute-force" search may be needed. In an interation where this condition is detected of the form of `(= ?r (conj b ?a))`, I think all subsequently dependent constraints that refer to `?r` need to be added ot the `constraints` immediately and removed from `more` - in terms of the locals used in the changes to the `non-equality-unifications` fn.

Reply:
Regarding the a-rule2 case you describe above @mrrodriguez it looks to me like the problem is that the alpha node is creating the ?r binding, but the join filter is then rebinding it rather than testing for equality with the value created in the alpha network.

````
 [5 :alpha-expr]
 [#object[clara.test_rules$eval17454$fn__17457 0x2fe5411 "clara.test_rules$eval17454$fn__17457@2fe5411"]
  {:file "*cider-repl localhost*",
   :compile-ctx
   {:condition
    {:type clara.test_rules.B,
     :constraints [(= ?r #{:b :a})],
     :original-constraints [(= ?r (conj b ?a)) (= ?r #{:b :a})]},
    :env nil,
    :msg "compiling alpha node"},
   :alpha-expr
   (clojure.core/fn
    [?__fact__ ?__env__]
    (clojure.core/let
     [this ?__fact__ ?__bindings__ (clojure.core/atom {})]
     (clojure.core/let
      [?r #{:b :a}]
      (clojure.core/swap! ?__bindings__ clojure.core/assoc :?r ?r)
      @?__bindings__)))}],
 [2 :join-filter-expr]
[2 :join-filter-expr]
 [#object[clara.test_rules$eval17454$fn__17459 0x60678a61 "clara.test_rules$eval17454$fn__17459@60678a61"]
  {:compile-ctx
   {:condition
    {:type clara.test_rules.B,
     :constraints [(= ?r #{:b :a})],
     :original-constraints [(= ?r (conj b ?a)) (= ?r #{:b :a})]},
    :join-filter-expressions
    {:type clara.test_rules.B, :constraints [(= ?r (conj b ?a))]},
    :env nil,
    :msg "compiling expression join node"},
   :join-filter-expr
   (clojure.core/fn
    [?__token__ ?__fact__ ?__element-bindings__ ?__env__]
    (clojure.core/let
     [this
      ?__fact__
      b
      (.-b ?__fact__)
      ?a
      (clojure.core/-> ?__token__ :bindings :?a)
      ?r
      (get ?__element-bindings__ :?r)
      ?__bindings__
      (clojure.core/atom {})]
     (clojure.core/let
      [?r (conj b ?a)]
      (clojure.core/swap! ?__bindings__ clojure.core/assoc :?r ?r)
      @?__bindings__)))}]
````

[compile-constraints](https://github.com/cerner/clara-rules/blob/eb8cf88cbae6d4493948e87a57c393dadb130b89/src/main/clojure/clara/rules/compiler.clj#L246) takes a collection of variables that it should not rebind, but in [compile-join-filter this only includes variables from the ancestors](https://github.com/cerner/clara-rules/blob/eb8cf88cbae6d4493948e87a57c393dadb130b89/src/main/clojure/clara/rules/compiler.clj#L500).  At first glance the solution would seem to be to pass along the element-bindings passed to compile-join-filter to compile-constraints.  However, on closer examination, it doesn't appear to me that the element-bindings are necessarily correct here.  They are [populated from the :new-bindings of the beta network](https://github.com/cerner/clara-rules/blob/eb8cf88cbae6d4493948e87a57c393dadb130b89/src/main/clojure/clara/rules/compiler.clj#L1361), which are just the bindings in the condition that aren't from the ancestors as can be seen in their population [here](https://github.com/cerner/clara-rules/blob/eb8cf88cbae6d4493948e87a57c393dadb130b89/src/main/clojure/clara/rules/compiler.clj#L928).

In order to fix this problem I think we need to store (correct) information on what bindings are created in the alpha network and which are created in the beta network, rather than just new-bindings information.  

Regarding this immediate PR, my current thinking is that there shouldn't be any need to start from the viewpoint of all bindings which are created in the condition; we can start from a "blank slate" of no bindings and proceed from there.  We could then generate a list of bindings created in the alpha network and of bindings created in the beta network and use this to resolve this issue and the one described by @mrrodriguez above.  Intuitively, the current solution seems a bit like we have incorrect binding analysis and are "cleaning it up" after the fact, while I'd suggest the cleaner solution would be to correct the binding analysis upfront.  Note also that we do have the list of bindings from the ancestors available should it be needed, although at the moment I don't think we do.  I also haven't tried implementing this concretely so it's possible that I'm missing something.

This may also bring up issues of whether we want the constraints to be order-independent as well and figuring out what the current behaviour is.